Update initial draft (#1)
privacysandbox/draft-ietf-protected-audience-key-value-service: draft-ietf-protected-audience-key-value-services.md

---
coding: utf-8

title: Protected Audience Key Value Services
abbrev: "KV Services"
docname: draft-ietf-protected-audience-key-value-services-latest
category: std
submissionType: IETF

area: TBD
workgroup: TBD
keyword: Internet-Draft

ipr: trust200902

stand_alone: yes
pi: [toc, sortrefs, symrefs]

author:
 -
    name: Peiwen Hu
    organization: Google
    email: peiwenhu@google.com
 -
    ins: B. R. Hamilton
    name: Benjamin "Russ" Hamilton
    organization: Google
    email: behamilton@google.com

normative:
  CBOR: RFC8949
  CDDL: RFC8610
  JSON: RFC8259
  OHTTP: RFC9458
  HPKE: RFC9180
  GZIP: RFC1952
  Brotli: RFC7932

informative:

--- abstract

The Key Value Service provides real-time signals to ad auctions while preserving user privacy.

--- middle

# Introduction

> [Protected Audience](https://wicg.github.io/turtledove/) is a privacy advancing API that facilitates
> interest group based advertising.

Reviewer: Words  from https://developers.google.com/privacy-sandbox/private-advertising/protected-audience
```suggestion
[Protected Audience](https://wicg.github.io/turtledove/) is a privacy advancing API that serves
remarketing and custom audiences use cases.
```

> Key Value Services are TEE based Key/Value databases that can be used to store

Reviewer: I think we need to define TEE before we use the term.

and integrate real-time data into Protected Audiences Auctions. The Protected
Audience proposal leverages Key Value Services to incorporate real-time signals

> into ad selection for both buyers and sellers. These servers provide a

Reviewer: ```suggestion
into ad selection for both buyers and sellers. These services provide a
```

flexible mechanism for fetching and processing data. While event-level logging

> is explicitly prohibited, the servers may have operational side effects like

Reviewer: ```suggestion
is explicitly prohibited, the services may have operational side effects like
```

use to communicate with the Key Value Service as part of the client's implementation of the
[Protected Audience API](https://wicg.github.io/turtledove/).

This document does not describe distribution of private keys to the Key Value Service.

## Terminology

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT",
"RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as
described in BCP 14 {{!RFC2119}} {{!RFC8174}} when, and only when, they appear in all capitals, as
shown here.

The key word "client" is to be interpreted as an implementation of this document that creates
Requests ({{request}}) and consumes Responses ({{response}}). The key phrase "Key Value Service" is
to be interpreted as an implementation of this document that consumes Requests and creates
Responses.

# Message Format Specifications {#format}

## Overview

To understand this document, it is important to know that the
communication between the client and the remote services uses a
request-response message exchange pattern.
On a high level, these request and response messages adhere to the following communication protocol:

> -   Data is transmitted over HTTPS using the `POST` method.

Reviewer: I think we should link HTTPS to the appropriate specification

The higher 6 bits are currently unused.

The following 4 bytes are the length of the request message in network byte order.

Then the request is zero padded to a set of pre-configured lengths.
Padding is applied with sizes as multiples of 2^n KBs ranging from 0 to 2MB. So the valid response
sizes will be `[0, 128B, 256B, 512B, 1KB, 2KB, 4KB, 8KB, 16KB, 32KB, 64KB, 128KB, 256KB, 512KB, 1MB,
2MB]`.

### Format+compression byte {#compression}

| Compression | Description                    |
| :---------: | :----------------------------- |
|    0x00     | [CBOR], no compression         |
|    0x01     | [CBOR], compressed in [Brotli] |
|    0x02     | [CBOR], compressed in [GZIP]   |

For requests, the byte value is 0x00. For responses, the byte value depends on the
`acceptCompression` (see {{request-schema}}) field in the request and the server behavior.

## Core Request Data {#request}

Requests are not compressed and have a tree-like hierarchy:

-   Each request contains one or more partitions. Each partition is a collection of keys that can be
    processed together by the service without any potential privacy leakage. Keys from one interest

>     group must be in the same partition. Keys from different interest groups with the same joining

Reviewer: I think we need to introduce the term interest group before using it.  Perhaps describe it as "an identifier of information about a user collected on a particular website or application".

    site (see https://wicg.github.io/turtledove/#joining-interest-groups) may or may not be in the same partition.
-   Each partition contains one or more key groups. Each key group has its unique attributes among
    all key groups in the partition. The attributes are represented by a list of `tags`. Besides

>     tags, the key group contains a list of keys to look up.

Reviewer: I find this really hard to understand.  I don't understand how each key group is "unique".  What is unique about it? is this a requirement of the key group or just something the client does?
I don't understand what the "tags" are for.  I think we need some motivation.
I don't understand what it means "to look up" a key.  Can we explain what this means and what it is looked up in?

Author: Changed the wording.

> -   Each partition has a unique id.

Reviewer: Why?  How is this used?

    partitions: [* partition],
    ; A list of partitions. Each must be processed independently. Accessible by UDF.
}

partition = {
    id: uint,
    ; Unique id of the partition in this request. Used by responses to refer to request partitions.
    compressionGroupId: uint,
    ; Unique id of a compression group in this request. Only partitions belonging to the same compression group will be compressed together in the response
    ? metadata: partitionMetadata,
    arguments: [* requestArgument],
    ; One group of keys and common attributes about them
}
;Single partition object. A collection of keys that can be processed together.

partitionMetadata = {
    ? hostname: tstr,
    ; The hostname of the top-level frame calling runAdAuction
    ? experimentGroupId: tstr,
    ? slotSize: tstr,
    ; Available if trustedBiddingSignalsSlotSizeMode=slot-size. In the form of <width>,<height>
    ? allSlotsRequestedSizes: tstr,
    ; Available if trustedBiddingSignalsSlotSizeMode=all-slots-requested-sizes. In the form of <width1>,<height1>,<width2>,<height2>,...
}

requestArgument = {
    ? tags: [* tstr],

>     ; List of tags describing this group's attributes

Reviewer: Are these picked from the below list of "Available Tags"?  If so, we should link to that section.
If this is an enumeration, we should declare its possible values rather than use the type "tstr" I think like Russ mentioned above for `acceptCompression`

Author: @peiwenhu are we prescribing that it is from the list of available tags? I don't think we should, since this will be input to the UDF.

Author: Discussed offline, we don't really want to restrict it to only those tags.

    ? data: [* tstr],
    ; List of keys to get values for
}
~~~~~

#### Available Tags

Each key group has exactly one tag from the `namespace` category.

| Tag | Description |
|---|---|
| interestGroupNames | Names of interest groups in the encompassing partition. |

> | keys | “keys” is a list of trustedBiddingSignalsKeys strings. |

Reviewer: trustedBiddingSignalsKeys needs a definition, which is going to be very complicated.  Perhaps it's better to give some more general description like `"keys" represent the keys to be used to look up values in a table` or something that ties back to the "to look up" mention above.

Author: Done.

> | renderUrls | Similarly, sellers may want to fetch information about a specific creative, e.g. the results of some out-of-band ad scanning system. This works in much the same way, with the base URL coming from the trustedScoringSignalsUrl property of the seller's auction configuration object.  |

Reviewer: I don't know what "Similarly" is referring similarity to.
It's unclear how "sellers may want to fetch information about a specific creative" relates to renderUrls.
trustedScoringSignalsUrl needs a definiition, which is going to be very complicated.  Perhaps its better to give a more general description like `"renderUrls" represent URLs for advertisements that are used to look up values in a table`

> | adComponentRenderUrls | Similar to renderUrls |

Reviewer: This needs more description.

---
coding: utf-8

title: Protected Audience Key Value Services
abbrev: "KV Services"
docname: draft-ietf-protected-audience-key-value-services-latest
category: std
submissionType: IETF

area: TBD
workgroup: TBD
keyword: Internet-Draft

ipr: trust200902

stand_alone: yes
pi: [toc, sortrefs, symrefs]

author:
 -
    name: Peiwen Hu
    organization: Google
    email: peiwenhu@google.com
 -

>     ins: B. R. Hamilton

Reviewer 2: I don't feel I contributed enough to be listed as an author here. Lusa, do you want to put yourself here?

Author: Sure, I'll add Tianyang too.

    name: Benjamin "Russ" Hamilton
    organization: Google
    email: behamilton@google.com

normative:
  CBOR: RFC8949
  CDDL: RFC8610
  JSON: RFC8259
  OHTTP: RFC9458
  HPKE: RFC9180
  GZIP: RFC1952
  Brotli: RFC7932

informative:

--- abstract

> The Key Value Service provides real-time signals to ad auctions while preserving user privacy.

Reviewer 2: What are real-time signals? What are ad auctions? The abstract can be longer than a sentence and should provide a description of what the standard is about that can be understood without any context.

Reviewer: +1 to what Russ said, perhaps something like "This document specifies a protocol for a service that can serve data with low latency and no side effects.  The data served can be used by clients for advertisement selection and the lack of side effects can be used to advance user privacy."

Author: Done.

--- middle

# Introduction

> Protected Audience is a privacy advancing API that facilitates

Reviewer: I feel like we should link to the Protected Audience spec

interest group based advertising.
Key Value Services are used by Protected Audience to provide

> real-time signals to buyers and sellers during ad selection.

Reviewer: ```suggestion
real-time signals to buyers and sellers during auctions performing ad selection.
```

> real-time signals to buyers and sellers during auctions performing ad selection.

Reviewer 2: define real-time signals. What sort of use cases (rate limiting, creative approval, etc)?

Author: Done.

The Protected Audience proposal leverages Key Value
services to incorporate real-time signals into ad selection.
These services utilize User-Defined Functions (UDFs) to
provide a flexible mechanism for fetching and processing data.
While event-level logging is explicitly prohibited, the servers
may have operational side effects like monitoring to ensure
security and prevent abuse.

## Scope

This document provides a specification for the request and response message format that a client can
use to communicate with the Key Value Service as part of the client's implementation of the

> Protected Audience API.

Reviewer: Should this also link to the PA spec?

Requests ({{request}}) and consumes Responses ({{response}}). The key phrase "Key Value Service" is
to be interpreted as an implementation of this document that consumes Requests and creates
Responses.

# Message Format Specifications {#format}

## Overview

To understand this document, it is important to know that the
communication between the client and the remote services uses a
request-response message exchange pattern.
On a high level, these request and response messages adhere to the following communication protocol:

-   Data is transmitted over HTTPS using the `POST` method.
-   Data within the request and response is encrypted with [HPKE].
-   The core request and response data is encoded using [CBOR].

### Encryption {#encryption}

The Key Value Service uses [HPKE] with the following configuration for encryption:

-   KEM (Key encapsulation mechanisms): 0x0020 DHKEM(X25519, HKDF-SHA256), see
    {{Section 1 of HPKE}}
-   KDF (key derivation functions): 0x0001 HKDF-SHA256, see {{Section 7.2 of HPKE}}
-   AEAD (Authenticated Encryption with Associated Data): AES-256-GCM, see
    {{Section 7.3 of HPKE}}

> The server is repurposing the [OHTTP] encapsulation mechanism ({{Section 4.6 of OHTTP}}), new

Reviewer 2: The definition of media types here is confusing. It might be better to move the description of media types to where they are actually used for the Request and Response. Then you can demonstrate how they are intended to be used.

Author: Done.

media types need to be defined:

-   The OHTTP request media type is `message/ad-auction-trusted-signals-request`
-   The OHTTP response media type is `message/ad-auction-trusted-signals-response`

These media types are concatenated with other fields when creating the [HPKE] encryption context.

### Message Framing {#framing}

Before encryption and after decryption, the requests and responses have the following framing:

| Byte     | 0         | 0             | 1 to 4   | 5 to Size+4       | Size+5 to end   |
| -------- | --------- | ------------- | -------- | ----------------- | --------------- |
| Bits     | 7-2       | 1-0           | \*       | \*                | \*              |
| -------- | --------- | ------------- | -------- | ----------------- | --------------- |
| Contents | Unused    | Compression   | Size     | Request Payload   | Padding         |

The request/response is framed with this 5 byte header.

The first byte is the format+compression byte. The lower 2 bits specify the format and compression ({{compression}}).
The higher 6 bits are currently unused.

The following 4 bytes are the length of the request message in network byte order.

> Then the request is zero padded to a set of pre-configured lengths.

Reviewer 2: "a set of pre-configured lengths" - What are these lengths?
Also, you may want to delegate descriptions of padding for requests and responses to subsections in the description of requests and responses.

Author: Moved to other section.

> Padding is applied with sizes as multiples of 2^n KBs ranging from 0 to 2MB. So the valid response

Reviewer 2: >Padding is applied with sizes as multiples of 2^n KBs ranging from 0 to 2MB.

This is for the response?

Author: Moved it to different sections so it's clearer.

sizes will be `[0, 128B, 256B, 512B, 1KB, 2KB, 4KB, 8KB, 16KB, 32KB, 64KB, 128KB, 256KB, 512KB, 1MB,

> 2MB]`.

Reviewer 2: What happens when the response doesn't fit in 2MB?

Author: @peiwenhu how do we want to handle large responses?

### Format+compression byte {#compression}

| Compression | Description                    |
| :---------: | :----------------------------- |
|    0x00     | [CBOR], no compression         |
|    0x01     | [CBOR], compressed in [Brotli] |
|    0x02     | [CBOR], compressed in [GZIP]   |

> 

Reviewer 2: ```
|    0x03     | Reserved                            |
```

Author: Done.

> For requests, the byte value is 0x00. For responses, the byte value depends on the

Reviewer 2: > For requests, the byte value is 0x00

Does that mean that requests are always uncompressed? Can we just say that - "Requests are always uncompressed so the Format+compression byte is 0x00"

Author: Done.

`acceptCompression` (see {{request-schema}}) field in the request and the server behavior.

> ## Core Request Data {#request}

Reviewer 2: What do you mean by "Core" here?

Author: Removed that

Requests are not compressed and have a tree-like hierarchy:

> -   Each request contains one or more partitions. Each partition is a collection of keys that can be

Reviewer 2: >Each partition is a collection of keys that can be processed together

Probably should be "Each partition is a collection of keys that SHALL be processed together"

We also probably don't need "without any potential privacy leakage". We may also want to add "Keys from separate partitions MUST NOT be processed together by the service." Then we could add a reference to the Security Considerations section where we can say that "Using separate partitions allow clients to break up the request to achieve desired privacy goals, such as preventing keys from an interest group joined on one origin from influencing the response for keys from an interest group joined on another origin".

>     processed together by the service without any potential privacy leakage. Keys from one interest

Reviewer 2: > Keys from one interest
    group must be in the same partition. Keys from different interest groups with the same joining
    site (see https://wicg.github.io/turtledove/#joining-interest-groups) may or may not be in the same partition.

I'm not sure this is necessary. It should be up to the client to decide how to partition keys when it assembles a request

Author: Done.

    group must be in the same partition. Keys from different interest groups with the same joining
    site (see https://wicg.github.io/turtledove/#joining-interest-groups) may or may not be in the same partition.
-   Each partition contains one or more key groups. Each key group has its unique attributes among
    all key groups in the partition. The attributes are represented by a list of `tags`. Besides
    tags, the key group contains a list of keys to look up.
-   Each partition has a unique id.
-   Each partition has a compression group field. Results of partitions belonging to the same
    compression group can be compressed together in the response. Different compression groups must

>     be compressed separately (see {{compression-group}}).

Reviewer 2: We could also add a reference here to the Security Considerations section describing why we need separate compression groups.

Author: Will do.

### Encryption {#request-encryption}

> The request is encrypted with [HPKE]. The request uses a repurposed [OHTTP] encapsulation format

Reviewer 2: Can we be more explicit here? I'd rather err on the side of being too explicit. In the B&A spec I wrote the procedure for encryption in a similar section to this.

Author: Done.

(see {{Section 4.3 of OHTTP}} and {{Section 4.6 of OHTTP}}) with a
`message/ad-auction-trusted-signals-request` media type instead of `message/bhttp request`.

### Request Schema {#request-schema}

The request is a [CBOR] encoded message with the following [CDDL] schema:

~~~~~ cddl
request = {

>     ? acceptCompression: [* tstr],

Reviewer 2: Instead of `tstr` as the type you should use a new type defined as `"none" / "gzip" / "brotli"`. That makes it explicit that this is a string enum.

Author: Done.

>     ; must contain at least one of "none", "gzip", "brotli"

Reviewer: ```suggestion
    ; A list of supported response compression algorithms; must contain at least one of "none", "gzip", "brotli"
```

    partitions: [* partition],
    ; A list of partitions. Each must be processed independently. Accessible by UDF.
}

partition = {
    id: uint,

>     ; Unique id of the partition in this request

Reviewer: ```suggestion
    ; Unique id of the partition in this request. Used by responses to refer to request partitions.
```

    compressionGroupId: uint,
    ; Unique id of a compression group in this request. Only partitions belonging to the same compression group will be compressed together in the response
    ? metadata: partitionMetadata,
    arguments: [* requestArgument],
    ; One group of keys and common attributes about them
}
;Single partition object. A collection of keys that can be processed together.

partitionMetadata = {
    ? hostname: tstr,
    ; The hostname of the top-level frame calling runAdAuction
    ? experimentGroupId: tstr,
    ? slotSize: tstr,

>     ; Available if trustedBiddingSignalsSlotSizeMode=slot-size. In the form of <width>,<height>

Reviewer 2: What is `trustedBiddingSignalsSlotSizeMode`? It's not defined in this document. Either omit it or reference the field in the Protected Audience spec (https://wicg.github.io/turtledove/#dom-generatebidinterestgroup-trustedbiddingsignalsslotsizemode)

Reviewer 2: That also seems to get into how a client might use the protocol instead of a description of the protocol.

Author: Done.

| Tag | Description |
|---|---|
| interestGroupNames | Names of interest groups in the encompassing partition. |
| keys | “keys” is a list of trustedBiddingSignalsKeys strings. |
| renderUrls | Similarly, sellers may want to fetch information about a specific creative, e.g. the results of some out-of-band ad scanning system. This works in much the same way, with the base URL coming from the trustedScoringSignalsUrl property of the seller's auction configuration object.  |
| adComponentRenderUrls | Similar to renderUrls |

### Generating a Request

TODO

### Parsing a Request {#request-parsing}

This section describes how the Key Value Service MUST deserialize request messages from the client.

The algorithm takes as input a serialized request message from the client and a list of HPKE private
keys (along with their corresponding key IDs).

The output is either an error sent back to the client, an empty message sent back to the client, or
a request message the Key Value Service can consume along with an HPKE context.

1. Let `encrypted request` be the request received from the client.
1. Let `error_msg` be an empty string.

> 1. De-encapsulate and decrypt `encrypted request` by using the input private key corresponding to

Reviewer 2: The method for de-encapsulation and decryption are not specified in this spec. You should describe in more detail either here or in the relevant section.

Reviewer 2: By de-encapsulation and decryption you seem to just mean decrypting not removing the framing. We should probably be more explicit about this.

Author: Done.

>    `key_id` as descripten in {{request-encryption}}, to get the decrypted message and `rctxt`.

Reviewer 2: descripten -> described

Author: Done.

    1. If decapsulation or decryption fails, return failure.
    1. Else, save the decrypted output as `framed request` and save `rctxt`.
1. Remove and extract the first 5 bytes from `framed request` as the `framing header` (described in
   {{framing}}), removing them from `framed request`.

>     1. If the `framing header`'s `Compression` field is not supported, return failure.

Reviewer 2: For requests the only supported compression is "none", correct?

Author: Right, fixed it.

1. Let `length` be equal to the `framing header`'s `Size` field.
1. If `length` is greater than the length of the remaining bytes in `framed request`, return
   failure.
1. Take the first `length` remaining bytes in `framed response` as `decodable request`, discarding
   the rest.

> 1. [CBOR] decode `decodable request` into the message represented in {{request-schema}}. Let this be

Reviewer 2: How do you handle decoding errors? Also, do you require deterministic CBOR? Are there any other requirements?

Author: Done.

   `processed request`.

> 1. Return `processed request` and `rctxt`.

Reviewer 2: What about handling fields? Are there scenarios where you would throw an error - like a compression group with no partitions, or too many partitions? What about missing fields?

### Example Request

> The [CBOR] representation consists of the following item, represented using the extended diagnostic

Reviewer 2: Let's emphasize that this is an example. "An example of the [CBOR] representation using the extended diagnostic notation from [CDDL] Appendix G" (also link "Appendix G" to the actual Appendix G here).

In general, examples should go in the Appendices since they are non-normative.

Author: Done.

notation from [CDDL] appendix G:

~~~~~ cbor-diag
{
  "acceptCompression": [
    "none",
    "gzip"
  ],
  "partitions": [
    {
      "id": 0,
      "compressionGroupId": 0,
      "metadata": {
        "hostname": "example.com",
        "experimentGroupId": "12345",
        "slotSize": "100,200",
      },
      "arguments": [
        {
          "tags": [
            "interestGroupNames"
          ],
          "data": [
            "InterestGroup1"
          ]
        },
        {
          "tags": [
            "keys"
          ],
          "data": [
            "keyAfromInterestGroup1",
            "keyBfromInterestGroup1"
          ]
        }
      ]
    },
    {
      "id": 1,
      "compressionGroupId": 0,
      "arguments": [
        {
          "tags": [
            "interestGroupNames"
          ],
          "data": [
            "InterestGroup2",
            "InterestGroup3"
          ]
        },
        {
          "tags": [
            "keys"
          ],
          "data": [
            "keyMfromInterestGroup2",
            "keyNfromInterestGroup3"
          ]
        }
      ]
    }
  ]
}
~~~~~

> ## Core Response Data {#response}

Reviewer 2: Again, I'm not sure what "Core" means here.

> The response is compressed. The compression is applied independently to each compression group. That

Reviewer 2: "The response is compressed" is not technically true for compression scheme "None".

Author: Done.

> The response is compressed. The compression is applied independently to each compression group. That

Reviewer 2: This paragraph feels like it belongs in {{response-schema}}.

Here we should probably have something describing a high level view of the responses - "An HPKE encrypted message sent as a Response to a Request, containing a framed top-level CBOR encoded payload that itself can contain multiple, possibly compressed, CBOR encoded messages."

Author: Done.

means, the response object mainly contains a list of compressed blobs, each for one compression
group. Each blob is for outputs of one or more partitions, sharing the same `compressionGroup` value
as specified in the request.

### Encryption {#response-encryption}

> The response uses the a similar encapsulated response format to that used by [OHTTP] (see

Reviewer 2: Be more explicit here as well.

Author: Done.

{{Section 4.4 of OHTTP}}), but with the custom `message/ad-auction-trusted-signals-response`
media type instead of `message/bhttp response`

### Response Schema {#response-schema}

The response is a [CBOR] encoded message with the following [CDDL] schema:

~~~~~ cddl
response = {
  ? compressionGroups : [* compressionGroup]
}

compressionGroup = {
  ? compressionGroupId: uint,
  ; Partition outputs with the same `compressionGroupId` specified in the request
  ; are compressed together.
  ? ttl_ms: uint,
  ; Adtech-specified TTL for client-side caching. In milliseconds. Unset means no caching.
  ? content: bstrs
  ; Compressed CBOR binary string using the algorithm specified in the request
  ; For details see compressed response content schema below.
}
~~~~~

##### CompressionGroup {#compression-group}

The content of each `compressionGroup` is a serialized [CBOR] list of partition outputs. This object
contains actual key value results for partitions in the corresponding compression group. The
uncompressed, deserialized [CBOR] content has the following [CDDL] schema:

~~~~~ cddl

> [* partitionOutput]

Reviewer 2: It's better to name things:
```
compressionGroup = [* partitionOutput]
```

Author: Done.

; Array of PartitionOutput objects

partitionOutput = {
  ? id: uint
  ; Unique id of the partition from the request
  ? keyGroupOutputs: [* keyGroupOutput]
}

keyGroupOutput = {
  tags: [* tstr]
  ; List of tags describing this key group's attributes
  ? keyValues: {
    ; At least one key-value pair if present
    * tstr => keyValue
  }
  ; One value to be returned in response for one key
  ; If a keyValues object exists, it must at least contain one key-value pair. If no key-value pair can be returned, the key group should not be in the response
}

keyValue = {

>   value: tstr

Reviewer 2: This is more of a question about the response format than the spec itself. Why do we have a dictionary containing a single string field instead of just putting the string field directly? It make be good to have a comment about this here in case someone asks.

Author: @peiwenhu I don't remember exactly why we had it like this. Do you?

}

~~~~~

> #### Example Compression Group

Reviewer 2: Move examples to Appendices.

Author: Done.

The [CBOR] representation consists of the following item, represented
using the extended diagnostic notation from [CDDL] appendix G:

~~~~~ cbor-diag
[
  {
    "id": 0,
    "keyGroupOutputs": [
      {
        "tags": [
          "interestGroupNames"
        ],
        "keyValues": {
          "InterestGroup1": {
            "value": "{\"priorityVector\":{\"signal1\":1}}"
          }
        }
      },
      {
        "tags": [
          "keys"
        ],
        "keyValues": {
          "keyAfromInterestGroup1": {
            "value": "valueForA"
          },
          "keyBfromInterestGroup1": {
            "value":"[\"value1ForB\",\"value2ForB\"]"
          }
        }
      }
    ]
  }
]
~~~~~

### Structured keys response specification

> Structured keys are keys that the browser is aware of and the browser
> can use the response to do
> additional processing. The value of these keys must abide by the
> following schema for the browser to

Reviewer: ```suggestion
Structured keys are keys that the client is aware of and the client
can use the response to do
additional processing. The value of these keys must abide by the
following schema for the client to
```

https://json-schema.org. For values for keys
from the `interestGroupNames` namespace, they must conform to the
following schema, prior to being
serialized to string.

~~~~~ json
{
    "title": "tkv.response.v2.InterestGroupResponse",
    "description": "Format for value of keys in groups tagged 'interestGroupNames'",
    "type": "object",
    "additionalProperties": false,
    "properties": {
        "priorityVector": {
            "type": "object",
            "patternProperties": {
                ".*": {
                    "description": "signals",
                    "type": "number"
                }
            }
        },
        "updateIfOlderThanMs": {
            "description": "This optional field specifies that the interest group should be updated if the interest group hasn't been joined or updated in a duration of time exceeding `updateIfOlderThanMs` milliseconds. Updates that ended in failure, either parse or network failure, are not considered to increment the last update or join time. An `updateIfOlderThanMs` that's less than 10 minutes will be clamped to 10 minutes.",
            "type": "unsigned integer"
        }
    }
}
~~~~~

> Example:

Reviewer 2: Examples go in Appendices.

~~~~~ json
{
    "priorityVector": {

>         "signal1": 1,

Reviewer 2: Use a different prefix for one of them. That makes it clear that they don't have to be "signal#"

Author: Done.

        "signal2": 2
    },
    "updateIfOlderThanMs": 10000
}
~~~~~

### Generating a Response

This algorithm describes how the Key Value Service MAY generate a response to a request.

> The input is a `payload` corresponding to {{response-schema}} and the HPKE receiver context saved in

Reviewer 2: It might be more natural to start with key-value responses to one or more partitions and compression groups. It feels like we're starting from the half-way point instead of the beginning.

Author: Done.

{{request-parsing}}, `rctxt`.

The output is a `response` to be sent to a Client.

1. Let `cbor payload` equal the
   [deterministically encoded CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c)
   `payload`. Return an empty `response` on CBOR encoding failure.

> 1. Let `compressed payload` equal the [GZIP] compressed `cbor payload`, returning an empty

Reviewer 2: This is not correct. We can't compress the top-level response. That leaves us open to compression ratio side-channel attacks (something to also mention in security considerations).

Author: Thanks for catching that.

   `response` on compression failure.
1. Create a framed payload, as described in {{framing}}:
    1. Create a `framing header`.
    1. Set the `framing header` `Compression` to one of 2.
    1. Set the `framing header` `Size` to the size of `compressed payload`.
    1. Let `framed payload` equal the result of prepending the framing header to
       `compressed payload`.
    2. Padding MAY be added to `framing header`.
    3. Return an empty `response` on failure of any of the previous steps.
2. Let `response` equal the result of the encryption and encapsulation of `framed payload` with
   `rctxt`, as described in {{response-encryption}}. Return an empty `response` on failure.
3. Return `response`.

### Parsing a Response

TODO

> # IANA Considerations

Reviewer 2: Add Security Considerations section.

Also IANA Considerations will likely include the new HTTP media types to be used with HTTP "content-type" headers (if specified in this spec).

Author: Done. Will start a doc on that.

|    0x02     | [CBOR], compressed in [GZIP]   |
|    0x03     | Reserved                       |

Requests are always uncompressed so the Format+compression byte is 0x00.

For responses, the byte value depends on the `acceptCompression` field
(see {{request-schema}}) in the request and the server behavior.

## Request Data {#request}

Requests are not compressed and have a tree-like hierarchy:

-   Each request contains one or more partitions. Each partition is a collection of keys that SHALL
    be processed together by the service. Keys from separate partitions MUST NOT be processed together by the service.
-   Each partition contains one or more key groups.
    Each key group contains a list of `tags` set by the client.
    Each key group contains a list of keys to be looked up in the service's internal datastore.
-   Each partition has a unique identifier.
    This allows the client to match the request partition with the corresponding
    `partitionOutput` (see {{compression-group}}) in the response.
-   Each partition has a compression group field.
    Results of partitions belonging to the same compression group can be compressed together
    in the response.

>     Different compression groups must be compressed separately (see {{compression-group}}).

Reviewer 2: "Different compression groups must be compressed separately..." That's referring to the response, but from the context it could refer to the request.

So maybe something like:
"The responses for different compression groups will be compressed separately in the response (see {{compression-group}}). Compressing the different groups separately avoids leaking the similarity of responses for different groups."

Author: Done.

   `Key Identifier`, `HPKE KEM ID`, `HPKE KDF ID`, and `HPKE AEAD ID` in network
   byte order.
2. Build a sequence of bytes (`info`) by concatenating the ASCII-encoded string
   "message/ad-auction-trusted-signals-request", a zero byte, and `hdr`.
3. Create a sending HPKE context by invoking `SetupBaseS()`
   ([Section 5.1.1](https://rfc-editor.org/rfc/rfc9180#section-5.1.1) of [HPKE])
   with the public key of the receiver `pkR` and `info`. This yields the context
   `sctxt` and an encapsulation key `enc`.
4. Encrypt `request` by invoking the `Seal()` method on `sctxt`
   ([Section 5.2](https://rfc-editor.org/rfc/rfc9180#section-5.2) of [HPKE])
   with empty associated data `aad`, yielding ciphertext `ct`.
5. Concatenate the values of `hdr`, `enc`, and `ct`.

In pseudocode, this procedure is as follows:

~~~~~
hdr = concat(encode(1, key_id),
             encode(2, kem_id),
             encode(2, kdf_id),
             encode(2, aead_id))
info = concat(encode_str("message/ad-auction-trusted-signals-request"),
              encode(1, 0),
              hdr)
enc, sctxt = SetupBaseS(pkR, info)
ct = sctxt.Seal("", request)
enc_request = concat(hdr, enc, ct)
~~~~~

> 

Reviewer 2: The client needs to save `sctxt`.

Author: Done.

   `key_id` as an 8-bit integer; plus
   `kem_id`, `kdf_id`, and `aead_id` as three 16-bit integers.
4. Create a receiving HPKE context, `rctxt`, by invoking `SetupBaseR()`
   ([Section 5.1.1](https://rfc-editor.org/rfc/rfc9180#section-5.1.1) of [HPKE])
   with `skR`, `enc`, and `info`.
5. Decrypt `ct` by invoking the `Open()` method on `rctxt`
   ([Section 5.2](https://rfc-editor.org/rfc/rfc9180#section-5.2) of [HPKE]),
   with an empty associated data `aad`, yielding `request` and returning an
   error on failure.

In pseudocode, this procedure is as follows:

~~~~~
key_id, kem_id, kdf_id, aead_id, enc, ct = parse(enc_request)
if version != 0 then return error
info = concat(encode_str("message/ad-auction-trusted-signals-request"),
              encode(1, 0),
              encode(1, key_id),
              encode(2, kem_id),
              encode(2, kdf_id),
              encode(2, aead_id))
rctxt = SetupBaseR(enc, skR, info)
request, error = rctxt.Open("", ct)
~~~~~

Key Value Services retain the HPKE context, `rctxt`, so that it can
encapsulate a response.

### Framing and Padding {#request-framing}

The plaintext request message uses the framing described in {{framing}}.

> Messages MAY be zero padded to a set of preconfigured lengths.

Reviewer 2: Are you going to specify the "preconfigured lengths" here? If not, you might just want to say that "Messages MAY be zero padded".

Author: Done.

    ? metadata: requestMetadata,
    partitions: [* partition],
    ; A list of partitions. Each must be processed independently. Accessible by UDF.
}

requestMetadata = {
    ? hostname: tstr,
    ; The hostname of the top-level frame calling runAdAuction
}

partition = {
    id: uint,
    ; Unique id of the partition in this request. Used by responses to refer to request partitions.
    compressionGroupId: uint,
    ; Unique id of a compression group in this request. Only partitions belonging to the same compression group will be compressed together in the response
    ? metadata: partitionMetadata,
    arguments: [* requestArgument],
    ; One group of keys and common attributes about them
}
;Single partition object. A collection of keys that can be processed together.

partitionMetadata = {
    ? experimentGroupId: tstr,
    ? slotSize: tstr,
    ? allSlotsRequestedSizes: tstr,
}

requestArgument = {
    ? tags: [* tstr],
    ; List of tags describing this group's attributes. These MAY be picked from the list of available tags in {{tags}}.
    ? data: [* tstr],
    ; List of keys to get values for
}
~~~~~

#### Available Tags {#tags}

> Each key group has exactly one tag from the `namespace` category.

Reviewer 2: It sounds like there are different types of tags: `namespace` tags and other unspecified tags. Can you describe this more explicitly?

Author: It looks like we've updated the API spec to drop namespace. Updated it.

a request message the Key Value Service can consume along with an HPKE context.

1. Let `encrypted request` be the request received from the client.
1. Let `error_msg` be an empty string.
1. Decrypt `encrypted request` by using the input private key corresponding to
   `key_id` as described in {{request-encryption}}, to get the decrypted message and `rctxt`.
    1. If decryption fails, return failure.
    2. Else, save the decrypted output as `framed request` and save `rctxt`.
2. Remove and extract the first 5 bytes from `framed request` as the `framing header` (described in
   {{framing}}), removing them from `framed request`.
    1. If the `framing header`'s `Compression` field is not `0x00` (no compression), return failure.
3. Let `length` be equal to the `framing header`'s `Size` field.
4. If `length` is greater than the length of the remaining bytes in `framed request`, return
   failure.
5. Take the first `length` remaining bytes in `framed response` as `decodable request`, discarding
   the rest.
6. [CBOR] decode `decodable request` into the message represented in {{request-schema}}. Let this be
   `processed request`.
    1. If decoding fails, return failure.

> 7. If no `partitions` are present, return failure.

Reviewer 2: Should the schema specify that at least one "partitions" is required (e.g. `partitions: [1* partition]`)?

8. Set `compression group map` to an empty map.
9. For each `partition` in `partitions`:

>    1. Set `partition id` to the value of the `id` field in `partition` if

Reviewer 2: In the schema, the `id` and `compressionGroupId` fields are required in the request.  It seems unusual to check for the presence of a required field.

Author: Done.

      present, otherwise `0`.

>    2. Set `compression group id` to the value of the `compressionGroupid` field in `partition` if

Reviewer 2: `compressionGroupid` -> `compressionGroupId`

Author: Done.

   `enc_response`. Note that `response_nonce` is of fixed length, so there is no
  ambiguity in parsing either `response_nonce` or `ct`.

In pseudocode, this procedure is as follows:

~~~~~
secret = context.Export("message/ad-auction-trusted-signals-response", max(Nn, Nk))
response_nonce = random(max(Nn, Nk))
salt = concat(enc, response_nonce)
prk = Extract(salt, secret)
aead_key = Expand(prk, "key", Nk)
aead_nonce = Expand(prk, "nonce", Nn)
ct = Seal(aead_key, aead_nonce, "", response)
enc_response = concat(response_nonce, ct)
~~~~~

Clients decrypt an Encapsulated Response by reversing this process. That is,
Clients first parse `enc_response` into `response_nonce` and `ct`. Then, they
follow the same process to derive values for `aead_key` and `aead_nonce`, using
their sending HPKE context, `sctxt`, as the HPKE context, `context`.

The Client uses these values to decrypt `ct` using the AEAD function `Open`.
Decrypting might produce an error, as follows:

~~~~~
response, error = Open(aead_key, aead_nonce, "", ct)
~~~~~

### Framing and Padding {#response-framing}

The plaintext response message uses the framing described in {{framing}}.

> Padding is applied with sizes as multiples of 2^n KBs ranging from 0 to 2MB. So the valid response

Reviewer 2: What does the server do if the response does not fit in 2MB?

Author: @peiwenhu

Reviewer 3: It returns an error response code. in the future the response code can be inside the ciphertext once we have error handling defined.

group. Each blob is for outputs of one or more partitions, sharing the same `compressionGroup` value
as specified in the request.

The response is a [CBOR] encoded message with the following [CDDL] schema:

~~~~~ cddl
response = {
  ? compressionGroups : [* compressionGroup]
}

compressionGroup = {
  ? compressionGroupId: uint,
  ; Partition outputs with the same `compressionGroupId` specified in the request
  ; are compressed together.
  ? ttl_ms: uint,
  ; Adtech-specified TTL for client-side caching. In milliseconds. Unset means no caching.
  ? content: bstrs
  ; Compressed CBOR binary string using the algorithm specified in the request
  ; For details see compressed response content schema below.
}
~~~~~

#### CompressionGroup {#compression-group}

The content of each `compressionGroup` is a serialized [CBOR] list of partition outputs. This object
contains actual key value results for partitions in the corresponding compression group. The
uncompressed, deserialized [CBOR] content has the following [CDDL] schema:

~~~~~ cddl
compressionGroup = [* partitionOutput]
; Array of PartitionOutput objects

partitionOutput = {

>   ? id: uint

Reviewer 2: Should this field be required? I'm not sure what a partitionOutput without an ID would mean.

Author: Done.

            "type": "object",
            "patternProperties": {
                ".*": {
                    "description": "signals",
                    "type": "number"
                }
            }
        },
        "updateIfOlderThanMs": {
            "description": "This optional field specifies that the interest group should be updated if the interest group hasn't been joined or updated in a duration of time exceeding `updateIfOlderThanMs` milliseconds. Updates that ended in failure, either parse or network failure, are not considered to increment the last update or join time. An `updateIfOlderThanMs` that's less than 10 minutes will be clamped to 10 minutes.",
            "type": "unsigned integer"
        }
    }
}
~~~~~

### Generating a Response

This algorithm describes how the Key Value Service MAY generate a response to a request.

The input is a list of [deterministically encoded CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c) `partitionOutputs` in {{response-schema}} as well as
the `compression group map` and the HPKE receiver, `rctxt`, context saved in {{request-parsing}}.

The output is a `response` to be sent to a Client.

> 1. Create `payload`.

Reviewer 2: What data type is `payload`?

Author: Done.

2. Set `compression groups` to an empty list.
3. Set `partition output map` to an empty map.
4. For each `partitionOutput` in the list of `partitionOutputs`:
   1. Set `partition id` to `partitionOutput["id"]` if present, otherwise 0.

>    2. Add key value pair `partition id`, `partitionOutput` to `partition output map`.

Reviewer 2: Maybe clearer to rewrite as "Set `partition output map[partition id]` to `partitionOutput`. Also consider renaming `partition output map` to not have spaces - either `partition_output_map` or `partitionOutputMap`.

Author: This is technically not correct, actually. Keeping it as is.

5. For each (`compression group id`, `partition ids`) in `compression group map`:

>    1. Create `compression group`.

Reviewer 2: What is the type of `compression group`?

Author: Done.

   2. Set `cbor partitions array` to an empty CBOR array.
   3. For each `partition id` in `partition ids`:
      1. Set `partition output` to `partition output map[partition id]`.
         1. On failure to find the partition, continue.

>       2. Add `partition output` to `cbor partitions array`.

Reviewer 2: Add -> Append?

Author: Done.

   4. Set `cbor serialized payload` to the CBOR serialized `cbor partitions array`.
      1. On serialization failure, continue.

>    5. Set `compression group content` to the [GZIP] compressed `cbor serialized payload`.

Reviewer 2: It doesn't have to be GZIP though. Maybe we should specify at the top that this response is to a request that included "gzip" as one of the acceptable compressions.

Author: Done.

>       1. On failure, set `compression group content` to empty.

Reviewer 2: So on compression failure, we send an empty message field instead of omitting it entirely? It seems like we should just skip the entire compression group.

Author: Done.

   6. Set `compression group["compressionGroupId"]` to `compression group id`
   7. Set `compression group["content"]` to `compression group content`.
   8. Add `compression group` to `compression groups`.
1. Set `payload["compressionGroups"]` to `compression groups`.
2. Create a framed payload, as described in {{framing}}:
    1. Create a `framing header`.
    2. Set the `framing header` `Compression` to one of 2.
    3. Set the `framing header` `Size` to the size of `compressed payload`.
    4. Let `framed payload` equal the result of prepending the framing header to
       `payload`.

>     5. Padding MAY be added to `framing header`.

Reviewer 2: `framing header` -> `framed payload`.

Author: Done.

   `processed request`.
    1. If decoding fails, return failure.
7. If no `partitions` are present, return failure.
8. Set `compressionGroupMap` to an empty map.
9. For each `partition` in `partitions`:

>    1. Set `compressionGroupMap[compression group id]` to `compressionGroupMap`.

Reviewer 2: So the values in `compressionGroupMap` are...itself? Shouldn't it be a list of partitions?

Author: Thanks for catching that. The values should be a list of partition Ids.

This algorithm describes how the Key Value Service MAY generate a response to a request.

> The input is a list of [deterministically encoded CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c) `partitionOutputs` in {{response-schema}} as well as

Reviewer 2: You may want to specify how the Key Value Service creates the "partitionOutputs"
 - as the output from the UDF given the keys (and what other fields) in the corresponding input
 - describe constraints on UDF?

Author: Thanks, added a short paragraph on UDFs.
